domain/secret/service/watcher.go

```diff
@@ -51,44 +51,87 @@ func NewWatchableService(
+			ctx, unitName, localConsumedRevisionIDs...)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		remoteConsumedSecretIDs, err := queryRemote(ctx, runner)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		remoteSecretURIs, err := s.secretState.GetConsumedRemoteSecretURIsWithChanges(
+			ctx, unitName, remoteConsumedSecretIDs...)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		result := append(localSecretURIs, remoteSecretURIs...)
+		if len(result) == 0 {
+			return nil, nil
+		}
+		return result, nil
 	}
 
-	tableRemote, queryRemote := s.secretState.InitialWatchStatementForConsumedRemoteSecretsChange(unitName)
-	wRemote, err := s.watcherFactory.NewNamespaceWatcher(
-		// We are interested in both CREATE and UPDATE changes on secret_reference table.
-		queryRemote,
+	return s.watcherFactory.NewNamespaceMapperWatcher(
+		initialQuery,
+		consumedSecretsChangesMapperFunc(s.secretState, unitName, tableLocal,
+			tableRemote),
+		// We are only interested in CREATE changes because the
+		// secret_revision.revision is immutable anyway.
+		eventsource.NamespaceFilter(tableLocal, changestream.Changed),
+		// We are interested in both CREATE and UPDATE changes on
+		// secret_reference table.
 		eventsource.NamespaceFilter(tableRemote, changestream.All),
 	)
-	if err != nil {
-		return nil, errors.Capture(err)
-	}
-	processRemoteChanges := func(ctx context.Context, secretIDs ...string) ([]string, error) {
-		return s.secretState.GetConsumedRemoteSecretURIsWithChanges(ctx, unitName, secretIDs...)
-	}
-	sWRemote, err := newSecretStringWatcher(wRemote, s.logger, processRemoteChanges)
-	if err != nil {
-		return nil, errors.Capture(err)
+}
+
+func consumedSecretsChangesMapperFunc(
+	state State,
+	unitName coreunit.Name,
+	tableLocal, tableRemote string,
+) eventsource.Mapper {
+	return func(ctx context.Context, changes []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		var localRevisionUUIDs, remoteSecretIDs []string
+		for _, change := range changes {
+			switch change.Namespace() {
+			case tableLocal:
+				localRevisionUUIDs = append(localRevisionUUIDs, change.Changed())
+			case tableRemote:
+				remoteSecretIDs = append(remoteSecretIDs, change.Changed())
+			}
+		}
+		var secretURIs []string
+		if len(localRevisionUUIDs) > 0 {
+			localSecretURIs, err := state.GetConsumedSecretURIsWithChanges(ctx, unitName, localRevisionUUIDs...)
+			if err != nil {
+				return nil, errors.Capture(err)
+			}
+			secretURIs = append(secretURIs, localSecretURIs...)
+		}
+		if len(remoteSecretIDs) > 0 {
+			remoteSecretURIs, err := state.GetConsumedRemoteSecretURIsWithChanges(ctx, unitName, remoteSecretIDs...)
+			if err != nil {
+				return nil, errors.Capture(err)
+			}
+			secretURIs = append(secretURIs, remoteSecretURIs...)
+		}
+		var result []changestream.ChangeEvent
+		for _, secretURI := range secretURIs {
+			result = append(result, secretURIChange(secretURI))
```

Comment:
We can just use the exisitng newMaskedChangeIDEvent(e, secretURI) which only mutates the ID but keeps the namespace and type, etc.